### ISSUE TITLE
could not initialize audio

### ISSUE THREAD

**Issue by ballastabwerfen** [supercollider/supercollider#4463]
## Environment

* SuperCollider version: 3.10.2
* Operating system: Windows 10 1803
* Other details (Qt version, audio driver, etc.): see below

## Steps to reproduce

```SuperDirt.start;
```

## Expected behaviour
Audio works. 

## Actual behavior
```
SC_PortAudioDriver: PortAudio failed at Pa_OpenStream with error: 'Unanticipated host error'
could not initialize audio.
SCDoc: Indexed 1917 documents in 2.62 seconds
Server 'localhost' exited with exit code -1073740791.
```
Yes I know there is a threat covering the exact same problem, 
[https://github.com/supercollider/supercollider/issues/3976](https://github.com/supercollider/supercollider/issues/3976)
but no I cannot reproduce the steps to solve this issue.

**SuperCollider seems to be working fine.**
```
*** Welcome to SuperCollider 3.10.2. *** For help press Ctrl-D.
SCDoc: Indexing help-files...
SCDoc: Indexed 1917 documents in 3.77 seconds
```

**Running the SuperDirt startup command.**

```
SuperDirt.start;
```

**This is the result.**
```
-> SuperDirt
Booting server 'localhost' on address 127.0.0.1:57110.

Device options:
  - MME : Microsoft Soundmapper - Input   (device #0 with 2 ins 0 outs)
  - MME : Mikrofonarray (Realtek High Def   (device #1 with 2 ins 0 outs)
  - MME : Microsoft Soundmapper - Output   (device #2 with 0 ins 2 outs)
  - MME : Lautsprecher (Realtek High Defi   (device #3 with 0 ins 6 outs)
  - Windows DirectSound : Primärer Soundaufnahmetreiber   (device #4 with 2 ins 0 outs)
  - Windows DirectSound : Mikrofonarray (Realtek High Definition Audio(SST))   (device #5 with 2 ins 0 outs)
  - Windows DirectSound : Primärer Soundtreiber   (device #6 with 0 ins 2 outs)
  - Windows DirectSound : Lautsprecher (Realtek High Definition Audio(SST))   (device #7 with 0 ins 6 outs)
  - ASIO : ASIO4ALL v2   (device #8 with 4 ins 12 outs)
  - Windows WASAPI : Lautsprecher (Realtek High Definition Audio(SST))   (device #9 with 0 ins 2 outs)
  - Windows WASAPI : Mikrofonarray (Realtek High Definition Audio(SST))   (device #10 with 2 ins 0 outs)
  - Windows WDM-KS : Headphones 1 (Realtek HD Audio 2nd output with SST)   (device #11 with 0 ins 2 outs)
  - Windows WDM-KS : Headphones 2 (Realtek HD Audio 2nd output with SST)   (device #12 with 0 ins 6 outs)
  - Windows WDM-KS : PC-Lautsprecher (Realtek HD Audio 2nd output with SST)   (device #13 with 2 ins 0 outs)
  - Windows WDM-KS : Headset Microphone (Realtek HD Audio Mic input)   (device #14 with 2 ins 0 outs)
  - Windows WDM-KS : Speakers 1 (Realtek HD Audio output with SST)   (device #15 with 0 ins 2 outs)
  - Windows WDM-KS : Speakers 2 (Realtek HD Audio output with SST)   (device #16 with 0 ins 6 outs)
  - Windows WDM-KS : PC-Lautsprecher (Realtek HD Audio output with SST)   (device #17 with 2 ins 0 outs)
  - Windows WDM-KS : Mikrofonarray (Realtek HD Audio Mic Array input)   (device #18 with 2 ins 0 outs)
  - Windows WDM-KS : Kopfhörer (@System32\drivers\bthhfenum.sys,#2;%1 Hands-Free%0


;(Bose Mini II SoundLink))   (device #19 with 0 ins 1 outs)
  - Windows WDM-KS : Kopfhörermikrofon (@System32\drivers\bthhfenum.sys,#2;%1 Hands-Free%0


;(Bose Mini II SoundLink))   (device #20 with 1 ins 0 outs)
  - Windows WDM-KS : Lautsprecher ()   (device #21 with 0 ins 2 outs)

Booting with:
  In: MME : Mikrofonarray (Realtek High Def
  Out: MME : Lautsprecher (Realtek High Defi
SC_PortAudioDriver: PortAudio failed at Pa_OpenStream with error: 'Unanticipated host error'
could not initialize audio.
Server 'localhost' exited with exit code -1073740791.
```

I took it from this post [https://github.com/supercollider/supercollider/issues/3976](https://github.com/supercollider/supercollider/issues/3976) 
that this behaviour is not a bug, but rather an issue with PortAudio and my specific system, 
yet following the threat mentioned above, the solution to this problem, basically switching audio devices upon startup, simply doesn't work for me, hence the bug report. 

**So I tried this:**

```
s.options.outDevice = "ASIO : ASIO4ALL v2   (device #8 with 4 ins 12 outs)";
s.options.inDevice = "ASIO : ASIO4ALL v2   (device #8 with 4 ins 12 outs)";
s.boot;
```

**This is the outcome:**
```
-> a ServerOptions
-> a ServerOptions
-> localhost
Booting server 'localhost' on address 127.0.0.1:57110.

Device options:
  - MME : Microsoft Soundmapper - Input   (device #0 with 2 ins 0 outs)
  - MME : Mikrofonarray (Realtek High Def   (device #1 with 2 ins 0 outs)
  - MME : Microsoft Soundmapper - Output   (device #2 with 0 ins 2 outs)
  - MME : Lautsprecher (Realtek High Defi   (device #3 with 0 ins 6 outs)
  - Windows DirectSound : Primärer Soundaufnahmetreiber   (device #4 with 2 ins 0 outs)
  - Windows DirectSound : Mikrofonarray (Realtek High Definition Audio(SST))   (device #5 with 2 ins 0 outs)
  - Windows DirectSound : Primärer Soundtreiber   (device #6 with 0 ins 2 outs)
  - Windows DirectSound : Lautsprecher (Realtek High Definition Audio(SST))   (device #7 with 0 ins 6 outs)
  - ASIO : ASIO4ALL v2   (device #8 with 4 ins 12 outs)
  - Windows WASAPI : Lautsprecher (Realtek High Definition Audio(SST))   (device #9 with 0 ins 2 outs)
  - Windows WASAPI : Mikrofonarray (Realtek High Definition Audio(SST))   (device #10 with 2 ins 0 outs)
  - Windows WDM-KS : Headphones 1 (Realtek HD Audio 2nd output with SST)   (device #11 with 0 ins 2 outs)
  - Windows WDM-KS : Headphones 2 (Realtek HD Audio 2nd output with SST)   (device #12 with 0 ins 6 outs)
  - Windows WDM-KS : PC-Lautsprecher (Realtek HD Audio 2nd output with SST)   (device #13 with 2 ins 0 outs)
  - Windows WDM-KS : Headset Microphone (Realtek HD Audio Mic input)   (device #14 with 2 ins 0 outs)
  - Windows WDM-KS : Speakers 1 (Realtek HD Audio output with SST)   (device #15 with 0 ins 2 outs)
  - Windows WDM-KS : Speakers 2 (Realtek HD Audio output with SST)   (device #16 with 0 ins 6 outs)
  - Windows WDM-KS : PC-Lautsprecher (Realtek HD Audio output with SST)   (device #17 with 2 ins 0 outs)
  - Windows WDM-KS : Mikrofonarray (Realtek HD Audio Mic Array input)   (device #18 with 2 ins 0 outs)
  - Windows WDM-KS : Kopfhörer (@System32\drivers\bthhfenum.sys,#2;%1 Hands-Free%0


;(Bose Mini II SoundLink))   (device #19 with 0 ins 1 outs)
  - Windows WDM-KS : Kopfhörermikrofon (@System32\drivers\bthhfenum.sys,#2;%1 Hands-Free%0


;(Bose Mini II SoundLink))   (device #20 with 1 ins 0 outs)
  - Windows WDM-KS : Lautsprecher ()   (device #21 with 0 ins 2 outs)

Booting with:
  In: MME : Mikrofonarray (Realtek High Def
  Out: MME : Lautsprecher (Realtek High Defi
SC_PortAudioDriver: PortAudio failed at Pa_OpenStream with error: 'Unanticipated host error'
could not initialize audio.
Server 'localhost' exited with exit code -1073740791.
```
It didn't switch the device. 
According to -> a ServerOptions something happened, but it didn't effect the outcome. 
As a matter of fact I tried all sorts of different variations of this code to rule out possible syntax hiccups.
The other thread mentioned leaving out spaces, but it was very unspecific at that part. 
Trying to switch to different devices other than ASIO4ALL yields the same result, it simply doesn't switch away from MME. 
Even having read the other threat a few times I am kind of stuck here and can't get the audio to work. 
I am not familiar with SuperCollider and I got little to no experience in programming at all, so this could easily be a user inflicted problem.
Maybe I am to dumb to run the code properly, or I am messing with the proper order of things I have to run to get a decent result and 
I am just not getting it because I am new to SuperCollider.
As far as my knowledge goes, at this point I cannot rule out the possibility of a bug in the interpretation of s.options.outDevince that's why I post this.
But I am actually expecting someone to write: "You got to push the red button twice." and everything works fine, or something.

Thanks in advance.

**Comment by jrsurge** [comment]
Hi!

Thanks for the detailed report.

Can you try changing the device selection code to:

```supercollider
s.options.device = "ASIO : ASIO 4ALL";
```

and report back if that helps?

The text in the brackets isn't actually part of the device name, just some extra information, so I don't think it's finding the device properly? And because it's ASIO, you can only use one device.

**Comment by jrsurge** [comment]
(I edited your submission to add a few more code sections with ` characters to show what came from the post window)

**Comment by jamshark70** [comment]
I believe it will be

```
s.options.device = "ASIO : ASIO4ALL v2";

// Or maybe this, but I think the first one
s.options.device = "ASIO : ASIO4ALL";
```

The earlier suggestion didn't catch that "v2" is (maybe) part of the name, and it inserted an extra space. The extra space will definitely cause the device name not to be recognized.

There's a valid enhancement request here, to make it clearer what is the formal device name in the scsynth output. Currently you just have to "know" where to stop copying, which is a kind of ridiculous position to put brand new users into.

**Comment by jrsurge** [comment]
>The extra space will definitely cause the device name not to be recognized. 

Yes! Absolutely, apologies (edited for clarity).

> The earlier suggestion didn't catch that "v2" is (maybe) part of the name, and it inserted an extra space. 

It can do a partial string match, so it should be fine (with the exception of the space, sorry again) - it just needs to be a unique match or it'll guess a device. If there's only one ASIO device, "ASIO" ought to work

**Comment by ballastabwerfen** [comment]
The good news is that both of the commands work.

```
s.options.device = "ASIO : ASIO4ALL v2";
```

and

```
s.options.device = "ASIO : ASIO4ALL";
```

The audio device switched to ASIO, the server started and seems to be fine. 
Nice!

```
Booting with:
  In: ASIO : ASIO4ALL v2
  Out: ASIO : ASIO4ALL v2
  Sample rate: 44100.000
  Latency (in/out): 0.016 / 0.016 sec
SC_AudioDriver: sample rate = 44100.000000, driver's block size = 64
SuperCollider 3 server ready.
Requested notification messages from server 'localhost'
localhost: server process's maxLogins (1) matches with my options.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
[...]
---- core synth defs loaded ----
[...]
loading 219 sample banks:
[...]
... file reading complete. Required 445 MB of memory.

SuperDirt: listening to Tidal on port 57120
server 'localhost' disconnected shared memory interface
Shared memory server interface initialized
```

Thank you very much for the very fast and profound help jrsurge and jamshark70!

**Comment by ballastabwerfen** [comment]
Addendum.

Following the steps of the [Start tidalcycles and superdirt for the first time](https://tidalcycles.org/index.php/Start_tidalcycles_and_superdirt_for_the_first_time) tutorial,
I realized that I still heard no audio. 

I figured out that the the problem was my Windows 10 audio driver settings.
For other audio purposes I run Windows audio driver regularly at 24bit 48000Hz.
Either ASIO4ALL or Supercollider don't support this, or I didn't configure them properly.
For the time being I resolved all issues setting the Windows audio driver back to 16bit 44100Hz.

This let me get rid of this:

```
SuperDirt: listening to Tidal on port 57120
server 'localhost' disconnected shared memory interface
Shared memory server interface initialized
```

And now indeed the final line of the SuperDirt startup is:

```
SuperDirt: listening to Tidal on port 57120
```

With this I could successfully complete the tutorial walkthrough.
